src/core/mac/mac.cpp

```diff
@@ -2257,39 +2257,52 @@ uint8_t Mac::GetTimeIeOffset(const Frame &aFrame)
 #endif
 
 #if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
-void Mac::SetCslChannel(uint8_t aChannel)
+void Mac::CslSample(void)
 {
-    VerifyOrExit(GetCslChannel() != aChannel);
+    bool     sample  = false;
+    uint16_t period  = Get<Mle::Mle>().IsRxOnWhenIdle() ? 0 : GetCslPeriod();
+    uint8_t  channel = GetCslChannel() ? GetCslChannel() : mRadioChannel;
 
-    mLinks.GetSubMac().SetCslChannel(aChannel);
-    mLinks.GetSubMac().SetCslChannelSpecified(aChannel != 0);
+    VerifyOrExit(IsCslEnabled() || (period == 0));
 
-    if (IsCslEnabled())
+    if (period != mLinks.GetSubMac().GetCslPeriod())
     {
-        Get<Mle::Mle>().ScheduleChildUpdateRequest();
+        mLinks.GetSubMac().SetCslPeriod(period);
+        IgnoreError(Get<Radio>().EnableCsl(mLinks.GetSubMac().GetCslPeriod(), Get<Mle::Mle>().GetParent().GetRloc16(),
+                                           &Get<Mle::Mle>().GetParent().GetExtAddress()));
+        sample = true;
     }
-exit:
-    return;
-}
 
-void Mac::SetCslPeriod(uint16_t aPeriod)
-{
-    mLinks.GetSubMac().SetCslPeriod(aPeriod);
-
-    Get<DataPollSender>().RecalculatePollPeriod();
-
-    if ((GetCslPeriod() == 0) || IsCslEnabled())
+    if (channel != mLinks.GetSubMac().GetCslChannel())
     {
-        IgnoreError(Get<Radio>().EnableCsl(GetCslPeriod(), Get<Mle::Mle>().GetParent().GetRloc16(),
-                                           &Get<Mle::Mle>().GetParent().GetExtAddress()));
+        mLinks.GetSubMac().SetCslChannel(channel);
+        sample = true;
     }
 
-    if (IsCslEnabled())
+    if (sample)
     {
-        Get<Mle::Mle>().ScheduleChildUpdateRequest();
+        mLinks.GetSubMac().CslSample();
+        if (period)
+        {
+            Get<Mle::Mle>().ScheduleChildUpdateRequest();
+        }
     }
 
-    UpdateIdleMode();
+exit:
+    return;
+}
+
+void Mac::SetCslChannel(uint8_t aChannel)
+{
+    mCslChannel = aChannel;
+    CslSample();
+}
+
+void Mac::SetCslPeriod(uint16_t aPeriod)
+{
+    mCslPeriod = aPeriod;
+    CslSample();
```

Comment:
Here is a flow actions which I think may be problematic with the changes in this PR:

- `SetCslPeriod()` can be called at any point of time (user can call `otLinkCslSetPeriod()` at any point).
- So it can be called while we are in middle of a frame tx (`SubMac` is in tx state).
- `SetCslPeriod()` will call the newly added `Mac::CslSample()`.
- From `Mac::CslSample()` we can end up calling `mLinks.GetSubMac().CslSample();` (e.g. if period is changed)
- `SubMac::CslSample()` then calls `SubMac::HandleCslTimer()`
- `HandleCslTimer()` will only look at csl state and may call `Get<Radio>().Sleep())` 
- So I guess we can end up putting radio to sleep in middle of an ongoing tx.